src/cycles/basis.jl

```diff
@@ -0,0 +1,72 @@
+# Code in this file inspired by NetworkX.
+
+"""
+    cycle_basis(g, root=nothing)
+
+Returns a list of cycles which form a basis for cycles of `g`.
+
+A basis for cycles of a network is a minimal collection of
+cycles such that any cycle in the network can be written
+as a sum of cycles in the basis.  Here summation of cycles
+is defined as "exclusive or" of the edges. Cycle bases are
+useful, e.g. when deriving equations for electric circuits
+using Kirchhoff's Laws.
+
+# Arguments
+- `g::AbstractGraph`: Graph
+- `root=nothing`: Specify a starting node for basis
+
+# Example
+```jldoctest
+julia> nodes = [1,2,3,4,5]
+julia> edges = [(1,2),(2,3),(3,4),(4,1),(1,5)]
+julia> g = Graph(length(nodes))
+julia> for e in edges add_edge!(G, e) end
+julia> cycle_basis(g)::Union{Int64, Void}
+2-element Array{Array{Int64,1},1}:
+ [2, 3, 4]
+ [2, 1, 3]
+```
+
+# References
+* Paton, K. An algorithm for finding a fundamental set of cycles of a graph. Comm. ACM 12, 9 (Sept 1969), 514-518. [https://dl.acm.org/citation.cfm?id=363232]
+"""
+function cycle_basis(g::AbstractGraph, root=nothing)
+    gnodes = Set(vertices(g))
+    cycles = Array{eltype(g),1}[]
+    while !isempty(gnodes)
+        if root == nothing
+            root = pop!(gnodes)
+        end
+        stack = [root]
+        pred = Dict(root => root)
+        used = Dict(root => [])
```

Comment:
I have got a general question about this: could the nodes also have non-integer names, e.g. a string "a"? I am not sure about this.

Reply:
you're absolutely right, that's just me not thinking enough ;)

Reply:
Nodes are currently always (1-indexed, consecutive) integers.

Reply:
Ok, so `SparseVector` might actually be a suitable alternative, esp. wrt the `key()` issue @sbromberger mentioned below. I'll check.

Reply:
this PR can go on with `Dict` for the moment and we'll see if changing to `SparseArrays` is worth it afterwards, maybe in another PR, this will also need to check if it is an actual improvement

Reply:
Well, hold on a sec. If one of you commits to exploring this once it's merged (and in the next couple of weeks), I'm ok with that. I really dislike Dicts :)

What say you?

Reply:
I could look into it. Maybe not next week but the week after.